src/sagas/assignments.js

```diff
@@ -0,0 +1,66 @@
+import {all, call, put, select, takeEvery} from 'redux-saga/effects';
+
+import {
+  getCourses,
+  createClassroomAssignment,
+} from '../clients/googleClassroom';
+import {
+  createProjectSnapshot,
+} from '../clients/firebase';
+import {
+  assignmentCreated,
+  assignmentNotCreated,
+} from '../actions/assignments';
+import {
+  coursesLoaded,
+} from '../actions/ui';
+import {
+  getCourse,
+  getCurrentProject,
+} from '../selectors';
+import {generateTextPreview} from '../util/compileProject';
+import {createSnapshotUrl} from '../util/exportUrls';
+
+export function* openAssignmentCreator() {
+  const courses = yield call(getCourses);
+  yield put(coursesLoaded(courses));
+}
+
+export function* createAssignment(
+  {payload: {selectedCourseId, dueDate, assignmentState}}) {
+  const project = yield select(getCurrentProject);
+  const snapshotKey = yield call(createProjectSnapshot, project);
+
+  const assignmentData = {
+    courseId: selectedCourseId,
+    dueDate,
+    url: yield call(createSnapshotUrl, snapshotKey),
+    title: yield call(generateTextPreview, project),
```

Comment:
I’m sure this works fine, but in a saga I think it’s a little easier to follow the flow if `yield`s are always in their own statement, in this case by assigning these to local variables above. I think you could also call `createProjectSnapshot` and `generateTextPreview` in parallel using the [`all` effect](https://redux-saga.js.org/docs/api/).